src/MSIdentityScaffolding/Microsoft.DotNet.MSIdentity/CodeReaderWriter/DocumentBuilder.cs

```diff
@@ -0,0 +1,533 @@
+            //adding usings
+            var usingNodes = CreateUsings(_codeFile.Usings);
+            if (usingNodes.Any() && _docRoot.Usings.Count == 0)
+            {
+                _documentEditor.ReplaceNode(_docRoot, _docRoot.WithUsings(new SyntaxList<UsingDirectiveSyntax>(usingNodes)));
+            }
+            else
+            {
+                foreach (var usingNode in usingNodes)
+                {
+                    //if usings exist (very likely), add it after the last one.
+                    if (_docRoot.Usings.Count > 0)
+                    {
+                        var usingName = usingNode.Name.ToString();
+                        if (!_docRoot.Usings.Any(node => node.Name.ToString().Equals(usingName)))
+                        {
+                            _documentEditor.InsertAfter(_docRoot.Usings.Last(), usingNode);
+                        }
+                    }
+                }
+            }
+        }
+
+        //Add class members to the top of the class.
+        public ClassDeclarationSyntax AddProperties(ClassDeclarationSyntax classDeclarationSyntax)
+        { 
+            var modifiedClassDeclarationSyntax = classDeclarationSyntax;
+            if (_codeFile.ClassProperties != null && _codeFile.ClassProperties.Any() && classDeclarationSyntax != null)
+            {
+                //get a sample member for leading trivia. Trailing trivia will still be semi colon and new line.
+                var sampleMember = modifiedClassDeclarationSyntax.Members.FirstOrDefault();
+                var memberLeadingTrivia = sampleMember?.GetLeadingTrivia() ?? new SyntaxTriviaList(SyntaxFactory.Whitespace("    "));
+                var memberTrailingTrivia = new SyntaxTriviaList(SemiColonTrivia, SyntaxFactory.CarriageReturnLineFeed);
+
+                //create MemberDeclarationSyntax[] with all the Property strings. 
+                var classProperties = CreateClassProperties(modifiedClassDeclarationSyntax.Members, memberLeadingTrivia, memberTrailingTrivia);
+
+                if (classProperties.Length > 0)
+                {
+                    //add to the top of the class.
+                    var members = modifiedClassDeclarationSyntax.Members;
+                    foreach (var property in classProperties)
+                    {
+                        members = members.Insert(0, property);
+                    }
+                    modifiedClassDeclarationSyntax = modifiedClassDeclarationSyntax
+                        .WithMembers(members);
+                }
+            }
+            return modifiedClassDeclarationSyntax;
+        }
+
+        //Add class attributes '[Attribute]' to a class.
+        public ClassDeclarationSyntax AddClassAttributes(ClassDeclarationSyntax classDeclarationSyntax)
+        {
+            var modifiedClassDeclarationSyntax = classDeclarationSyntax;
+            if (_codeFile.ClassAttributes != null && _codeFile.ClassAttributes.Any() && classDeclarationSyntax != null)
```

Comment:
>_codeFile.ClassAttributes != null && _codeFile.ClassAttributes.Any() [](http://example.com/codeflow?start=16&length=68)

same here if class already does not have attributes - we never add new ones?

Reply:
moved https://github.com/dotnet/Scaffolding/pull/1593